src/shogun/util/PutPerceptron_benchmark.cc

```diff
@@ -0,0 +1,137 @@
+#include <benchmark/benchmark.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/features/DotFeatures.h>
+#include <shogun/features/iterators/DotIterator.h>
+#include <shogun/labels/BinaryLabels.h>
+#include <shogun/mathematics/Math.h>
+
+namespace shogun
+{
+	// Mock PerceptronP
+	class PerceptronP : public CSGObject
+	{
+
+	public:
+		PerceptronP() : CSGObject()
+		{
+			max_iters = 1000;
+			learn_rate = 0.1;
+			SG_ADD(&w, "w", "weights", MS_AVAILABLE);
+		}
+
+		~PerceptronP()
+		{
+		}
+
+		bool train(CDotFeatures* feats, CBinaryLabels* labels, bool use_put)
+		{
+			int32_t num_feat = feats->get_dim_feature_space();
+			SGVector<float64_t> weights(num_feat);
+			weights.set_const(1.0 / num_feat);
+			bias = 0;
+
+			SGVector<float64_t> output(feats->get_num_vectors());
+			for (auto iter : range(max_iters))
+			{
+				auto iter_train_labels = labels->get_int_labels().begin();
+				auto iter_output = output.begin();
+
+				for (const auto& v : DotIterator(feats))
+				{
+					const auto true_label = *(iter_train_labels++);
+					auto& predicted_label = *(iter_output++);
+
+					predicted_label = v.dot(weights) + bias;
+
+					if (CMath::sign<float64_t>(predicted_label) != true_label)
+					{
+						const auto gradient = learn_rate * true_label;
+						bias += gradient;
+						v.add(gradient, weights);
+					}
+					if (use_put)
+					{
+						this->put<SGVector<float64_t>>("w", weights);
+					}
+					else
+						w = weights;
+				}
+			}
+		}
+
+		virtual const char* get_name() const
+		{
+			return "PerceptronP";
+		}
+
+	protected:
+		/** learning rate */
+		float64_t bias;
+		SGVector<float64_t> w;
+		float64_t learn_rate;
```

Comment:
i guess we can make learn rate to constant 1 and bias to constant 0 to simplify things